URIExtractionNamespace: Treat null values in lookup maps as missing entries.

gianm opened the pull request:
This is useful when many logical lookups are derived from the same base JSON file,
and some lookups' values may be unknown sometimes.

Inspired by https://groups.google.com/forum/#!topic/druid-user/_gcU2iW9O84.

drcrallen commented:
Have you tried using `""` for missing?

gianm commented:
@drcrallen not in the context of an actual extractionFn; is there a test that verifies that works as expected? If so then that's a good workaround.

Although even then, allowing `null` to mean "drop out of the lookup" seems like a nicer experience for users when the input json file may not be totally immaculate.

drcrallen commented:
So is the expected behavior of `null` to be "not included in lookup" or to be "map to missing"?

gianm commented:
I'd expect a null value or an empty string to both behave the same as if the row weren't there at all, and all three of those situations should be returned as `null` by the extractionFns.

drcrallen commented:
`io.druid.query.extraction.FunctionalExtractionTest` has various tests that ensure the `""` vs 'null' behavior is appropriate.

drcrallen commented:
I don't see one in that test suite for retain missing as true and using `PRESENT_VALUE`. I added it here: https://github.com/druid-io/druid/pull/3522

drcrallen commented on extensions-core/lookups-cached-global/src/main/java/io/druid/data/input/MapPopulator.java:
```diff
@@ -64,8 +64,11 @@ public long populate(final ByteSource source, final Map<K, V> map) throws IOExce
-            ++count;
+            final Map<K, V> kvMap = parser.parse(line);
+            if (kvMap != null) {
+              map.putAll(kvMap);
```
Not all maps can accept null values. This retains prior functionality of "potentially failing if target map can't handle null values", but I just wanted to make that callout here.

leventov replied:
@drcrallen `MapPopulator` is used in prod only once, in `URIExtractionNamespace `. There the value is checked to be not `null`. So it seems to be fine.

drcrallen commented:
Cool, @gianm can you please update docs for the extension to make a note that null values are now skipped?

b-slim commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/URIExtractionNamespace.java:
```diff
@@ -217,13 +217,12 @@ private DelegateParser(
+      final Object val = inner.get(value);
+      if (val == null) {
+        // Skip nulls
+        return null;
```
shall we do the same for empty string ?

leventov replied:
Since `Parser.parse()` is not marked as Nullable, maybe saner to return `emptyMap()` here and remove the check in `MapPopulator`?

gianm replied:
emptyMap sounds good, will change that.

b-slim commented:
LGTM not sure if we should do the check for empty as null as well.

leventov commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/URIExtractionNamespace.java:
```diff
@@ -217,13 +217,12 @@ private DelegateParser(
-      return ImmutableMap.<String, String>of(k, val);
+      final Object val = inner.get(value);
+      if (val == null) {
+        // Skip nulls
```
Could you please make the comment more descriptive, and add logging at least on debug level (or maybe higher)?

gianm replied:
I don't think there should be a debug message here, the whole point is that it could be normal for this to happen. But I will add a log.info message at the end of loading that calls out the number of lines read and entries parsed, so if those are unexpectedly different you can do something about it.

gianm commented:
@leventov @drcrallen @b-slim replaced null with empty map, added a test, and added more details to the post-populate log message.

gianm commented:
I also changed the lines counter to an int from a long, and added a check for overflow, because we can't store more than int.maxvalue entries in a map anyway.

gianm commented:
Also updated docs.

b-slim commented:
👍

leventov commented:
LGTM.

fjy commented:
👍
